lib/mayaUsd/nodes/layerManager.cpp

```diff
@@ -358,28 +375,45 @@ void LayerDatabase::prepareForWriteCheck(bool* retCode, bool isExport)
     }
 }
 
-bool LayerDatabase::getStagesToSave(bool isExport)
+bool LayerDatabase::getProxiesToSave(bool isExport)
 {
-    auto allStages = MayaUsd::ufe::getAllStages();
     bool checkSelection = isExport && (MFileIO::kExportTypeSelected == MFileIO::exportType());
     const UFE_NS::GlobalSelection::Ptr& ufeSelection = UFE_NS::GlobalSelection::get();
 
     _proxiesToSave.clear();
-    for (const auto& stage : allStages) {
-        auto stagePath = MayaUsd::ufe::stagePath(stage);
-        if (!stagePath.empty()) {
+    _internalProxiesToSave.clear();
+
+    MFnDependencyNode  fn;
+    MItDependencyNodes iter(MFn::kPluginDependNode);
+    for (; !iter.isDone(); iter.next()) {
+        MObject mobj = iter.item();
+        fn.setObject(mobj);
+        if (!fn.isFromReferencedFile() && supportedNodeType(fn.typeId())) {
+            MayaUsdProxyShapeBase* pShape = static_cast<MayaUsdProxyShapeBase*>(fn.userNode());
+            UsdStageRefPtr         stage = pShape ? pShape->getUsdStage() : nullptr;
+            if (!stage) {
+                continue;
+            }
+
+            auto stagePath = MayaUsd::ufe::stagePath(stage);
             if (!checkSelection
                 || (ufeSelection->contains(stagePath)
                     || ufeSelection->containsAncestor(stagePath))) {
-                // Remove the leading "|world| component.
-                std::string       strStagePath = stagePath.popHead().string();
-                MDagPath          proxyDagPath = UsdMayaUtil::nameToDagPath(strStagePath);
-                MFnDependencyNode fn(proxyDagPath.node());
-                if (!fn.isFromReferencedFile() && supportedNodeType(fn.typeId())) {
+                MPlug   shareStagePlug(mobj, MayaUsdProxyShapeBase::shareStageAttr);
+                bool    sharedStage = true;
+                MStatus status = shareStagePlug.getValue(sharedStage);
+                if ((status == MS::kSuccess && !sharedStage) || !pShape->isStageIncoming()) {
                     SdfLayerHandleVector allLayers = stage->GetLayerStack(true);
                     for (auto layer : allLayers) {
                         if (layer->IsDirty()) {
-                            _proxiesToSave.append(proxyDagPath);
+                            MDagPath proxyDagPath;
+                            MDagPath::getAPathTo(mobj, proxyDagPath);
+                            // We handle unshared compostion internally in Maya USD file
+                            if ((status == MS::kSuccess && sharedStage) || status != MS::kSuccess) {
```

Comment:
@ppt-adsk Yea that looks a bit confusing... 

Line 405: This predicate is "Should we save the stage?". 

1) shareStagePlug doesn't exist (older maya that doesn't have this) : In this case we case we only care about saving if the input is not an incoming connection, since in that case the node that "owns" the stage (upstream node) is responsible for saving. For example if you have multiple proxy shapes daisy chained one's out_stage feeding the other's in_stage. In that case only one proxy is responsible for saving (the first one). 

2) shareStagePlug exists: Similarly but in this case if the stage is unshared, it means we are responsible for saving the root layer (and stubs for the sublayers so we can put them back in the same spot). So doesn't matter if its incoming or not, we need to save. 

Line 412: This is more "Where should we save the stage?". The reason we have this distinction now is that some Layers are special case layers that should be saved to the maya file only. There are two examples currently of this, the session layer (was there before) and my newly added unshared root layer. So since we have batchSave and a delegate which handles saving externally, we need to manage some proxies ourselves and control where they save.. Ideally we should move all the code in one place but that's a bigger refactor probably. 

I've discussed some of these changes with @fowlertADSK

Reply:
O.K., sounds good, it would be great if you could convert a condensed version of your explanation above into a code comment.